### PR TITLE
[IMP] portal: add informative button to internal notes on portal chatter

### PR DESCRIPTION
**Pull request by pko-odoo**
PURPOSE

Before this commit, when the chatter was displayed on a portal
page one could see the internal notes with a button for toggling
his visibility... which was ineffective because a portal user
could never see internal notes and on the opposite internal users
could always see them.
This commit adds another informative (disabled) button for internal
notes.

SPECIFICATION

Adds a 'is_note' param on the message in the _portal_message_format
method.

LINKS

Task-2817892

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/portal/static/src/xml/portal_chatter.xml [hunk 1]
```diff
@@ -114,9 +114,11 @@
 
     <!-- Chatter: internal toggle widget -->
     <t t-name="portal.chatter_internal_toggle">
-        <div t-attf-class="float-right o_portal_chatter_js_is_internal #{message.is_internal and 'o_portal_message_internal_on' or 'o_portal_message_internal_off'}"
+        <div t-attf-class="float-right o_portal_chatter_js_is_internal #{message.is_note ? 'o_portal_message_subtype_internal' :
```
Comment (awa-odoo) [inline addons/portal/static/src/xml/portal_chatter.xml:117]: This becomes a bit complicated and hard to read, especially since there is JS code that toggles the classes "o_portal_message_internal_on" and "o_portal_message_internal_off".

Can't we instead:

```
<div t-if="message.is_note">
    -> your button here
</div>
<div t-else="" t-attf-class="..." ...> --> the current div and its content with the additional "t-else".
</div>
```

File: addons/portal/models/mail_message.py [hunk 2]
```diff
@@ -16,8 +16,11 @@ def portal_message_format(self):
 
     def _portal_message_format(self, fields_list):
         vals_list = self._message_format(fields_list)
+        note_id = self.env['ir.model.data'].xmlid_to_res_id('mail.mt_note')
         IrAttachmentSudo = self.env['ir.attachment'].sudo()
         for vals in vals_list:
+            message_sudo = self.browse(vals['id']).sudo().with_prefetch(self.ids)
+            vals['is_note'] = message_sudo.subtype_id.id == note_id
```
Comment (awa-odoo) [inline addons/portal/models/mail_message.py:23]: Is the subtype not available within vals?
Seems harsh to browse within a loop, even if we use the "with_prefetch" trick.

(I see you copy/pasted from other overrides of "message_format", but it seems like "subtype_id" is included in the vals if I look at the base docstring of message_format).

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

Left a couple comments to try to improve performances and readability.
In addition, it would be nice if your commit message explained that this is used for the "customer preview" feature of the portal.

Thanks & cheers!

File: addons/portal/models/mail_message.py [hunk 3]
```diff
@@ -16,8 +16,10 @@ def portal_message_format(self):
 
     def _portal_message_format(self, fields_list):
         vals_list = self._message_format(fields_list)
+        note_id = self.env['ir.model.data'].xmlid_to_res_id('mail.mt_note')
```
Comment (awa-odoo) [inline addons/portal/models/mail_message.py:19]: message_subtype_note_id ?

Side-note: not sure "xmlid_to_res_id" exists, isn't it "_xmlid_to_res_id" ?

File: addons/portal/models/mail_message.py [hunk 4]
```diff
@@ -16,8 +16,10 @@ def portal_message_format(self):
...
         IrAttachmentSudo = self.env['ir.attachment'].sudo()
         for vals in vals_list:
+            vals['is_note'] = vals['subtype_id'][0] == note_id
```
Comment (awa-odoo) [inline addons/portal/models/mail_message.py:22]: "is_message_subtype_note" ?

Also, just to be safe: `message_subtype_note_id and vals.get('subtype_id', [False])[0] == message_subtype_note_id` ?

File: addons/portal/static/src/scss/portal.scss [hunk 5]
```diff
@@ -510,14 +510,20 @@ form label {
...
     .o_portal_message_internal_off {
-        .btn-danger {
+        .btn-danger, .btn-secondary {
             display: none;
         }
     }
 
     .o_portal_message_internal_on {
-        .btn-success {
+        .btn-success, .btn-secondary {
```
Comment (awa-odoo) [inline addons/portal/static/src/scss/portal.scss:526]: Are these changes still necessary at all?

File: addons/portal/static/src/xml/portal_chatter.xml [hunk 6]
```diff
@@ -114,7 +114,10 @@
 
     <!-- Chatter: internal toggle widget -->
     <t t-name="portal.chatter_internal_toggle">
-        <div t-attf-class="float-right o_portal_chatter_js_is_internal #{message.is_internal and 'o_portal_message_internal_on' or 'o_portal_message_internal_off'}"
+        <div t-if="message.is_note" class="float-right o_portal_message_subtype_internal">
```
Comment (awa-odoo) [inline addons/portal/static/src/xml/portal_chatter.xml:117]: Not sure we need that extra class anymore.

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

Some more cleaning to do I believe ^^

Cheers!

File: addons/portal/models/mail_message.py (hunk 3 above)
Comment (pko-odoo) [reply addons/portal/models/mail_message.py]: It seems that there are some occurences of `_xmlid_to_res_id` in the code but there is no method definition of it :smiley:

File: addons/portal/models/mail_message.py [hunk 7]
```diff
@@ -16,8 +16,10 @@ def portal_message_format(self):
 
     def _portal_message_format(self, fields_list):
         vals_list = self._message_format(fields_list)
+        message_subtype_note_id = self.env['ir.model.data'].xmlid_to_res_id('mail.mt_note')
         IrAttachmentSudo = self.env['ir.attachment'].sudo()
         for vals in vals_list:
+            vals['is_message_subtype_note'] = message_subtype_note_id and vals.get('subtype_id', [False])[0] == message_subtype_note_id
```
Comment (awa-odoo) [inline addons/portal/models/mail_message.py:22]: Almost there, the commit message still mentions adding a "is_note" param ^^

As we made a lot of changes, not sure if this needs to be re-tested by Luc, I'm guessing you re-tested thoroughly?

**Comment by awa-odoo** [comment]
@robodoo r+
